Prepare to switch from std::variant to WTF::Variant

achristensen07 opened the pull request:
#### 7f62a88a674a810b02784b2d141a23839416215b
<pre>
Prepare to switch from std::variant to WTF::Variant
<a href="https://rdar.apple.com/149072846">rdar://149072846</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=291432">https://bugs.webkit.org/show_bug.cgi?id=291432</a>

Reviewed by Keith Miller.

libcpp&apos;s std::variant implementation generates suboptimal code.
See <a href="https://github.com/llvm/llvm-project/issues/62648">https://github.com/llvm/llvm-project/issues/62648</a>
We have partially mitigated this in <a href="https://commits.webkit.org/283806@main">https://commits.webkit.org/283806@main</a>
and <a href="https://commits.webkit.org/289736@main">https://commits.webkit.org/289736@main</a> by making our own std::visit
replacements, but that doesn&apos;t help std::variant&apos;s destructor or copy
constructor.

This patch adds a &lt;wtf/Variant.h&gt; header that simply includes &lt;variant&gt;.
This is step 1 of at least 3 to replace our use of std::variant.
Step 2 will be to add &quot;using Variant = std::variant&quot; and replace our
uses of std::variant with Variant, which will initially not change anything
because it will be just an alias, then step 3 will be to replace the alias
with our own implementation.

While I was at it, I put wtf/Variant.h into precompiled headers.  That
saved me from needing to realphabetize ~150 includes to meet WebKit style
and it also will speed up build time.

* Source/JavaScriptCore/JavaScriptCorePrefix.h:
* Source/JavaScriptCore/disassembler/Disassembler.cpp:
* Source/JavaScriptCore/interpreter/Interpreter.h:
* Source/JavaScriptCore/jit/AssemblyHelpers.h:
* Source/JavaScriptCore/jit/SnippetReg.h:
* Source/JavaScriptCore/parser/Lexer.cpp:
* Source/JavaScriptCore/parser/VariableEnvironment.h:
* Source/JavaScriptCore/runtime/BytecodeCacheError.h:
* Source/JavaScriptCore/runtime/CachePayload.h:
* Source/JavaScriptCore/runtime/CacheUpdate.h:
* Source/JavaScriptCore/runtime/TemporalObject.h:
* Source/JavaScriptCore/runtime/TemporalTimeZone.h:
* Source/JavaScriptCore/runtime/VM.h:
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/CompactVariantOperations.h:
* Source/WTF/wtf/CrossThreadCopier.h:
* Source/WTF/wtf/Expected.h:
* Source/WTF/wtf/FlatteningVariantAdaptor.h:
* Source/WTF/wtf/GenericHashKey.h:
* Source/WTF/wtf/Hasher.h:
* Source/WTF/wtf/LikelyDenseUnsignedIntegerSet.h:
* Source/WTF/wtf/SmallMap.h:
* Source/WTF/wtf/StdLibExtras.h:
* Source/WTF/wtf/Variant.h: Copied from Source/WebCore/workers/FetchingWorkerIdentifier.h.
* Source/WTF/wtf/VariantExtras.h:
* Source/WTF/wtf/VariantList.h:
* Source/WTF/wtf/text/TextBreakIterator.h:
* Source/WebCore/Modules/WebGPU/GPUBindGroupEntry.h:
* Source/WebCore/Modules/WebGPU/GPUColorDict.h:
* Source/WebCore/Modules/WebGPU/GPUError.h:
* Source/WebCore/Modules/WebGPU/GPUExtent3DDict.h:
* Source/WebCore/Modules/WebGPU/GPUImageCopyExternalImage.h:
* Source/WebCore/Modules/WebGPU/GPUOrigin2DDict.h:
* Source/WebCore/Modules/WebGPU/GPUOrigin3DDict.h:
* Source/WebCore/Modules/WebGPU/GPUPipelineDescriptorBase.h:
* Source/WebCore/Modules/WebGPU/GPURenderPassColorAttachment.h:
* Source/WebCore/Modules/WebGPU/GPURenderPassDepthStencilAttachment.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUBindGroupEntry.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUColor.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUError.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUExtent3D.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUImageCopyExternalImage.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUOrigin2D.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUOrigin3D.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPURenderPassColorAttachment.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPURenderPassDepthStencilAttachment.h:
* Source/WebCore/Modules/async-clipboard/ClipboardItemBindingsDataSource.h:
* Source/WebCore/Modules/fetch/FetchBody.h:
* Source/WebCore/Modules/fetch/FetchHeaders.h:
* Source/WebCore/Modules/indexeddb/IDBCursor.h:
* Source/WebCore/Modules/indexeddb/IDBGetAllResult.h:
* Source/WebCore/Modules/indexeddb/IDBKey.h:
* Source/WebCore/Modules/indexeddb/IDBKeyData.h:
* Source/WebCore/Modules/indexeddb/IDBKeyPath.h:
* Source/WebCore/Modules/indexeddb/IDBRequest.cpp:
* Source/WebCore/Modules/mediacontrols/MediaControlsHost.cpp:
* Source/WebCore/Modules/mediacontrols/MediaControlsHost.h:
* Source/WebCore/Modules/mediastream/MediaTrackConstraints.h:
* Source/WebCore/Modules/mediastream/RTCIceServer.h:
* Source/WebCore/Modules/paymentrequest/PaymentMethodChangeEvent.h:
* Source/WebCore/Modules/paymentrequest/PaymentRequest.h:
* Source/WebCore/Modules/push-api/PushEventInit.h:
* Source/WebCore/Modules/push-api/PushSubscription.h:
* Source/WebCore/Modules/push-api/PushSubscriptionOptionsInit.h:
* Source/WebCore/Modules/speech/SpeechRecognitionUpdate.h:
* Source/WebCore/Modules/webaudio/AudioContextOptions.h:
* Source/WebCore/Modules/webaudio/AudioNode.h:
* Source/WebCore/Modules/webcodecs/WebCodecsAudioDataAlgorithms.h:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoder.cpp:
* Source/WebCore/Modules/webxr/WebXRWebGLLayer.h:
* Source/WebCore/WebCorePrefix.h:
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXTreeStore.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.h:
* Source/WebCore/animation/AnimationEffect.h:
* Source/WebCore/animation/EffectTiming.h:
* Source/WebCore/animation/OptionalEffectTiming.h:
* Source/WebCore/bindings/IDLTypes.h:
* Source/WebCore/bindings/js/BufferSource.h:
* Source/WebCore/bindings/js/JSDOMConvertUnion.h:
* Source/WebCore/bindings/js/JSValueInWrappedObject.h:
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(AddToIncludesForIDLType):
* Source/WebCore/crypto/CryptoAlgorithm.h:
* Source/WebCore/crypto/CryptoKey.h:
* Source/WebCore/crypto/SubtleCrypto.h:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESKW.cpp:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmHMAC.cpp:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmRSASSA_PKCS1_v1_5.cpp:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmRSA_OAEP.cpp:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmRSA_PSS.cpp:
* Source/WebCore/crypto/parameters/CryptoAlgorithmEcdsaParams.h:
* Source/WebCore/crypto/parameters/CryptoAlgorithmHmacKeyParams.h:
* Source/WebCore/crypto/parameters/CryptoAlgorithmRsaHashedImportParams.h:
* Source/WebCore/crypto/parameters/CryptoAlgorithmRsaHashedKeyGenParams.h:
* Source/WebCore/crypto/parameters/CryptoAlgorithmX25519Params.h:
* Source/WebCore/css/CSSFontFaceSet.h:
* Source/WebCore/css/CSSStyleProperties.cpp:
* Source/WebCore/css/CSSStyleSheet.h:
* Source/WebCore/css/DOMMatrixReadOnly.h:
* Source/WebCore/css/FontFace.h:
* Source/WebCore/css/StyleRule.h:
* Source/WebCore/css/calc/CSSCalcTree.h:
* Source/WebCore/css/typedom/CSSNumericValue.h:
* Source/WebCore/css/typedom/CSSUnparsedValue.cpp:
* Source/WebCore/css/typedom/CSSUnparsedValue.h:
* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/css/values/color/CSSAbsoluteColorSerialization.h:
* Source/WebCore/css/values/color/CSSColor.h:
* Source/WebCore/css/values/color/CSSColorDescriptors.h:
* Source/WebCore/css/values/color/CSSRelativeColor.h:
* Source/WebCore/css/values/color/CSSRelativeColorSerialization.h:
* Source/WebCore/css/values/filter-effects/CSSFilterFunction.h:
* Source/WebCore/css/values/primitives/CSSUnevaluatedCalc.h:
* Source/WebCore/dom/ContainerNode.cpp:
* Source/WebCore/dom/DocumentMarker.h:
* Source/WebCore/dom/EventTarget.h:
* Source/WebCore/dom/MessageEvent.h:
* Source/WebCore/dom/Node.cpp:
* Source/WebCore/editing/AlternativeTextController.h:
* Source/WebCore/fileapi/Blob.h:
* Source/WebCore/fileapi/NetworkSendQueue.h:
* Source/WebCore/html/DOMFormData.h:
* Source/WebCore/html/HTMLAllCollection.cpp:
* Source/WebCore/html/ImageBitmap.cpp:
* Source/WebCore/html/ImageDataArray.h:
* Source/WebCore/html/URLSearchParams.h:
* Source/WebCore/html/canvas/CanvasPath.h:
* Source/WebCore/html/canvas/CanvasStyle.h:
* Source/WebCore/html/canvas/GPUCanvasContext.h:
* Source/WebCore/html/canvas/GPUCanvasContextCocoa.h:
* Source/WebCore/html/canvas/WebGLFramebuffer.h:
* Source/WebCore/inspector/InspectorCanvas.cpp:
* Source/WebCore/inspector/InspectorCanvas.h:
* Source/WebCore/inspector/InspectorCanvasCallTracer.cpp:
* Source/WebCore/inspector/agents/InspectorCanvasAgent.cpp:
* Source/WebCore/layout/integration/inline/InlineIteratorBox.h:
* Source/WebCore/layout/integration/inline/InlineIteratorLineBox.h:
* Source/WebCore/loader/ResourceLoaderOptions.h:
* Source/WebCore/page/DiagnosticLoggingClient.h:
* Source/WebCore/page/IntersectionObserver.h:
* Source/WebCore/page/LocalDOMWindow.cpp:
* Source/WebCore/page/Performance.h:
* Source/WebCore/page/PerformanceMeasureOptions.h:
* Source/WebCore/page/scrolling/ScrollingCoordinator.h:
* Source/WebCore/platform/Cursor.h:
* Source/WebCore/platform/PasteboardCustomData.h:
* Source/WebCore/platform/ProcessIdentity.h:
* Source/WebCore/platform/SharedBuffer.h:
* Source/WebCore/platform/audio/AudioStreamDescription.h:
* Source/WebCore/platform/calc/CalculationTree.h:
* Source/WebCore/platform/graphics/ColorInterpolationMethod.h:
* Source/WebCore/platform/graphics/Font.h:
* Source/WebCore/platform/graphics/FontCascadeDescription.h:
* Source/WebCore/platform/graphics/FontPalette.h:
* Source/WebCore/platform/graphics/FontPaletteValues.h:
* Source/WebCore/platform/graphics/FontSizeAdjust.h:
* Source/WebCore/platform/graphics/Gradient.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayerDelegatedContents.h:
* Source/WebCore/platform/graphics/cocoa/SourceBufferParser.h:
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.h:
* Source/WebCore/platform/graphics/cocoa/UnrealizedCoreTextFont.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListItem.h:
* Source/WebCore/platform/network/DNS.h:
* Source/WebCore/platform/network/FormData.h:
* Source/WebCore/platform/sql/SQLValue.h:
* Source/WebCore/platform/sql/SQLiteStatement.cpp:
* Source/WebCore/platform/text/TextFlags.h:
* Source/WebCore/platform/xr/PlatformXR.h:
* Source/WebCore/rendering/style/StyleGridData.h:
* Source/WebCore/rendering/svg/RenderSVGResourcePaintServer.h:
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/testing/TypeConversions.h:
* Source/WebCore/workers/FetchingWorkerIdentifier.h:
* Source/WebCore/workers/WorkerInspectorProxy.h:
* Source/WebCore/workers/service/ExtendableMessageEvent.h:
* Source/WebCore/workers/service/ServiceWorkerRoute.h:
* Source/WebCore/workers/service/ServiceWorkerTypes.h:
* Source/WebCore/xml/XMLHttpRequest.h:
* Source/WebGPU/WGSL/WGSL.h:
* Source/WebKit/GPUProcess/ShapeDetection/ShapeDetectionObjectHeap.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/WebGPUObjectHeap.h:
* Source/WebKit/NetworkProcess/NetworkLoadChecker.h:
* Source/WebKit/NetworkProcess/cache/NetworkCacheData.h:
* Source/WebKit/Platform/IPC/ArgumentCoders.h:
* Source/WebKit/Platform/IPC/MessageReceiveQueueMap.h:
* Source/WebKit/Platform/cocoa/MediaPlaybackTargetContextSerialized.h:
* Source/WebKit/Scripts/PreferencesTemplates/WebPreferencesStoreDefaultsMap.cpp.erb:
* Source/WebKit/Scripts/webkit/messages.py:
(class_template_headers):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebCompiledContentRuleListData.h:
* Source/WebKit/Shared/WebFoundTextRange.h:
* Source/WebKit/Shared/WebGPU/WebGPUColor.h:
* Source/WebKit/Shared/WebGPU/WebGPUError.h:
* Source/WebKit/Shared/WebGPU/WebGPUExtent3D.h:
* Source/WebKit/Shared/WebGPU/WebGPUImageCopyExternalImage.h:
* Source/WebKit/Shared/WebGPU/WebGPUOrigin2D.h:
* Source/WebKit/Shared/WebGPU/WebGPUOrigin3D.h:
* Source/WebKit/Shared/WebGPU/WebGPURenderPassColorAttachment.h:
* Source/WebKit/Shared/WebGPU/WebGPURenderPassDepthStencilAttachment.h:
* Source/WebKit/UIProcess/API/APIWebAuthenticationPanel.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Automation/SimulatedInputDispatcher.cpp:
* Source/WebKit/UIProcess/Automation/SimulatedInputDispatcher.h:
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/SubFrameSOAuthorizationSession.h:
* Source/WebKit/UIProcess/Cocoa/_WKWarningView.h:
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebAuthentication/WebAuthenticationRequestData.h:
* Source/WebKit/WebKit2Prefix.h:
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferBackendHandle.h:
* Source/WebKitLegacy/mac/DOM/DOMHTMLOptionsCollection.mm:
* Tools/TestWebKitAPI/Tests/WTF/CrossThreadCopierTests.cpp:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/FrameTreeChecks.h:

Canonical link: <a href="https://commits.webkit.org/293592@main">https://commits.webkit.org/293592@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3fb079bc0efb09e4a8785f307fffe2d036470a7f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99382 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19031 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9283 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104513 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/49983 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/101423 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19320 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27465 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/75639 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/49983 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102389 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/14693 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/89728 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/55998 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/98855 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/14487 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/7711 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49343 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/92065 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/84413 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/7798 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/106870 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/98001 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/26496 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/19328 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/106870 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/26858 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/85932 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/106870 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/6472 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/20239 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16169 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/26436 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/31637 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/121617 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/26256 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33967 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/29569 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/27823 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 reviewed (approved):
r=me but can we be a bit clearer on what steps this PR is making in the commit message? e.g. "This patch adds a `WTF/wtf/Variant.h` header that simply forwards to `<variant>`. This is step 1 of at least 3 to replace our use of std::variant."

achristensen07 committed:
Prepare to switch from std::variant to WTF::Variant
rdar://149072846
https://bugs.webkit.org/show_bug.cgi?id=291432

Reviewed by Keith Miller.

libcpp's std::variant implementation generates suboptimal code.
See https://github.com/llvm/llvm-project/issues/62648
We have partially mitigated this in https://commits.webkit.org/283806@main
and https://commits.webkit.org/289736@main by making our own std::visit
replacements, but that doesn't help std::variant's destructor or copy
constructor.

This patch adds a <wtf/Variant.h> header that simply includes <variant>.
This is step 1 of at least 3 to replace our use of std::variant.
Step 2 will be to add "using Variant = std::variant" and replace our
uses of std::variant with Variant, which will initially not change anything
because it will be just an alias, then step 3 will be to replace the alias
with our own implementation.

While I was at it, I put wtf/Variant.h into precompiled headers.  That
saved me from needing to realphabetize ~150 includes to meet WebKit style
and it also will speed up build time.

* Source/JavaScriptCore/JavaScriptCorePrefix.h:
* Source/JavaScriptCore/disassembler/Disassembler.cpp:
* Source/JavaScriptCore/interpreter/Interpreter.h:
* Source/JavaScriptCore/jit/AssemblyHelpers.h:
* Source/JavaScriptCore/jit/SnippetReg.h:
* Source/JavaScriptCore/parser/Lexer.cpp:
* Source/JavaScriptCore/parser/VariableEnvironment.h:
* Source/JavaScriptCore/runtime/BytecodeCacheError.h:
* Source/JavaScriptCore/runtime/CachePayload.h:
* Source/JavaScriptCore/runtime/CacheUpdate.h:
* Source/JavaScriptCore/runtime/TemporalObject.h:
* Source/JavaScriptCore/runtime/TemporalTimeZone.h:
* Source/JavaScriptCore/runtime/VM.h:
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/CompactVariantOperations.h:
* Source/WTF/wtf/CrossThreadCopier.h:
* Source/WTF/wtf/Expected.h:
* Source/WTF/wtf/FlatteningVariantAdaptor.h:
* Source/WTF/wtf/GenericHashKey.h:
* Source/WTF/wtf/Hasher.h:
* Source/WTF/wtf/LikelyDenseUnsignedIntegerSet.h:
* Source/WTF/wtf/SmallMap.h:
* Source/WTF/wtf/StdLibExtras.h:
* Source/WTF/wtf/Variant.h: Copied from Source/WebCore/workers/FetchingWorkerIdentifier.h.
* Source/WTF/wtf/VariantExtras.h:
* Source/WTF/wtf/VariantList.h:
* Source/WTF/wtf/text/TextBreakIterator.h:
* Source/WebCore/Modules/WebGPU/GPUBindGroupEntry.h:
* Source/WebCore/Modules/WebGPU/GPUColorDict.h:
* Source/WebCore/Modules/WebGPU/GPUError.h:
* Source/WebCore/Modules/WebGPU/GPUExtent3DDict.h:
* Source/WebCore/Modules/WebGPU/GPUImageCopyExternalImage.h:
* Source/WebCore/Modules/WebGPU/GPUOrigin2DDict.h:
* Source/WebCore/Modules/WebGPU/GPUOrigin3DDict.h:
* Source/WebCore/Modules/WebGPU/GPUPipelineDescriptorBase.h:
* Source/WebCore/Modules/WebGPU/GPURenderPassColorAttachment.h:
* Source/WebCore/Modules/WebGPU/GPURenderPassDepthStencilAttachment.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUBindGroupEntry.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUColor.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUError.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUExtent3D.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUImageCopyExternalImage.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUOrigin2D.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUOrigin3D.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPURenderPassColorAttachment.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPURenderPassDepthStencilAttachment.h:
* Source/WebCore/Modules/async-clipboard/ClipboardItemBindingsDataSource.h:
* Source/WebCore/Modules/fetch/FetchBody.h:
* Source/WebCore/Modules/fetch/FetchHeaders.h:
* Source/WebCore/Modules/indexeddb/IDBCursor.h:
* Source/WebCore/Modules/indexeddb/IDBGetAllResult.h:
* Source/WebCore/Modules/indexeddb/IDBKey.h:
* Source/WebCore/Modules/indexeddb/IDBKeyData.h:
* Source/WebCore/Modules/indexeddb/IDBKeyPath.h:
* Source/WebCore/Modules/indexeddb/IDBRequest.cpp:
* Source/WebCore/Modules/mediacontrols/MediaControlsHost.cpp:
* Source/WebCore/Modules/mediacontrols/MediaControlsHost.h:
* Source/WebCore/Modules/mediastream/MediaTrackConstraints.h:
* Source/WebCore/Modules/mediastream/RTCIceServer.h:
* Source/WebCore/Modules/paymentrequest/PaymentMethodChangeEvent.h:
* Source/WebCore/Modules/paymentrequest/PaymentRequest.h:
* Source/WebCore/Modules/push-api/PushEventInit.h:
* Source/WebCore/Modules/push-api/PushSubscription.h:
* Source/WebCore/Modules/push-api/PushSubscriptionOptionsInit.h:
* Source/WebCore/Modules/speech/SpeechRecognitionUpdate.h:
* Source/WebCore/Modules/webaudio/AudioContextOptions.h:
* Source/WebCore/Modules/webaudio/AudioNode.h:
* Source/WebCore/Modules/webcodecs/WebCodecsAudioDataAlgorithms.h:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoder.cpp:
* Source/WebCore/Modules/webxr/WebXRWebGLLayer.h:
* Source/WebCore/WebCorePrefix.h:
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXTreeStore.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.h:
* Source/WebCore/animation/AnimationEffect.h:
* Source/WebCore/animation/EffectTiming.h:
* Source/WebCore/animation/OptionalEffectTiming.h:
* Source/WebCore/bindings/IDLTypes.h:
* Source/WebCore/bindings/js/BufferSource.h:
* Source/WebCore/bindings/js/JSDOMConvertUnion.h:
* Source/WebCore/bindings/js/JSValueInWrappedObject.h:
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(AddToIncludesForIDLType):
* Source/WebCore/crypto/CryptoAlgorithm.h:
* Source/WebCore/crypto/CryptoKey.h:
* Source/WebCore/crypto/SubtleCrypto.h:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESKW.cpp:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmHMAC.cpp:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmRSASSA_PKCS1_v1_5.cpp:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmRSA_OAEP.cpp:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmRSA_PSS.cpp:
* Source/WebCore/crypto/parameters/CryptoAlgorithmEcdsaParams.h:
* Source/WebCore/crypto/parameters/CryptoAlgorithmHmacKeyParams.h:
* Source/WebCore/crypto/parameters/CryptoAlgorithmRsaHashedImportParams.h:
* Source/WebCore/crypto/parameters/CryptoAlgorithmRsaHashedKeyGenParams.h:
* Source/WebCore/crypto/parameters/CryptoAlgorithmX25519Params.h:
* Source/WebCore/css/CSSFontFaceSet.h:
* Source/WebCore/css/CSSStyleProperties.cpp:
* Source/WebCore/css/CSSStyleSheet.h:
* Source/WebCore/css/DOMMatrixReadOnly.h:
* Source/WebCore/css/FontFace.h:
* Source/WebCore/css/StyleRule.h:
* Source/WebCore/css/calc/CSSCalcTree.h:
* Source/WebCore/css/typedom/CSSNumericValue.h:
* Source/WebCore/css/typedom/CSSUnparsedValue.cpp:
* Source/WebCore/css/typedom/CSSUnparsedValue.h:
* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/css/values/color/CSSAbsoluteColorSerialization.h:
* Source/WebCore/css/values/color/CSSColor.h:
* Source/WebCore/css/values/color/CSSColorDescriptors.h:
* Source/WebCore/css/values/color/CSSRelativeColor.h:
* Source/WebCore/css/values/color/CSSRelativeColorSerialization.h:
* Source/WebCore/css/values/filter-effects/CSSFilterFunction.h:
* Source/WebCore/css/values/primitives/CSSUnevaluatedCalc.h:
* Source/WebCore/dom/ContainerNode.cpp:
* Source/WebCore/dom/DocumentMarker.h:
* Source/WebCore/dom/EventTarget.h:
* Source/WebCore/dom/MessageEvent.h:
* Source/WebCore/dom/Node.cpp:
* Source/WebCore/editing/AlternativeTextController.h:
* Source/WebCore/fileapi/Blob.h:
* Source/WebCore/fileapi/NetworkSendQueue.h:
* Source/WebCore/html/DOMFormData.h:
* Source/WebCore/html/HTMLAllCollection.cpp:
* Source/WebCore/html/ImageBitmap.cpp:
* Source/WebCore/html/ImageDataArray.h:
* Source/WebCore/html/URLSearchParams.h:
* Source/WebCore/html/canvas/CanvasPath.h:
* Source/WebCore/html/canvas/CanvasStyle.h:
* Source/WebCore/html/canvas/GPUCanvasContext.h:
* Source/WebCore/html/canvas/GPUCanvasContextCocoa.h:
* Source/WebCore/html/canvas/WebGLFramebuffer.h:
* Source/WebCore/inspector/InspectorCanvas.cpp:
* Source/WebCore/inspector/InspectorCanvas.h:
* Source/WebCore/inspector/InspectorCanvasCallTracer.cpp:
* Source/WebCore/inspector/agents/InspectorCanvasAgent.cpp:
* Source/WebCore/layout/integration/inline/InlineIteratorBox.h:
* Source/WebCore/layout/integration/inline/InlineIteratorLineBox.h:
* Source/WebCore/loader/ResourceLoaderOptions.h:
* Source/WebCore/page/DiagnosticLoggingClient.h:
* Source/WebCore/page/IntersectionObserver.h:
* Source/WebCore/page/LocalDOMWindow.cpp:
* Source/WebCore/page/Performance.h:
* Source/WebCore/page/PerformanceMeasureOptions.h:
* Source/WebCore/page/scrolling/ScrollingCoordinator.h:
* Source/WebCore/platform/Cursor.h:
* Source/WebCore/platform/PasteboardCustomData.h:
* Source/WebCore/platform/ProcessIdentity.h:
* Source/WebCore/platform/SharedBuffer.h:
* Source/WebCore/platform/audio/AudioStreamDescription.h:
* Source/WebCore/platform/calc/CalculationTree.h:
* Source/WebCore/platform/graphics/ColorInterpolationMethod.h:
* Source/WebCore/platform/graphics/Font.h:
* Source/WebCore/platform/graphics/FontCascadeDescription.h:
* Source/WebCore/platform/graphics/FontPalette.h:
* Source/WebCore/platform/graphics/FontPaletteValues.h:
* Source/WebCore/platform/graphics/FontSizeAdjust.h:
* Source/WebCore/platform/graphics/Gradient.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayerDelegatedContents.h:
* Source/WebCore/platform/graphics/cocoa/SourceBufferParser.h:
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.h:
* Source/WebCore/platform/graphics/cocoa/UnrealizedCoreTextFont.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListItem.h:
* Source/WebCore/platform/network/DNS.h:
* Source/WebCore/platform/network/FormData.h:
* Source/WebCore/platform/sql/SQLValue.h:
* Source/WebCore/platform/sql/SQLiteStatement.cpp:
* Source/WebCore/platform/text/TextFlags.h:
* Source/WebCore/platform/xr/PlatformXR.h:
* Source/WebCore/rendering/style/StyleGridData.h:
* Source/WebCore/rendering/svg/RenderSVGResourcePaintServer.h:
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/testing/TypeConversions.h:
* Source/WebCore/workers/FetchingWorkerIdentifier.h:
* Source/WebCore/workers/WorkerInspectorProxy.h:
* Source/WebCore/workers/service/ExtendableMessageEvent.h:
* Source/WebCore/workers/service/ServiceWorkerRoute.h:
* Source/WebCore/workers/service/ServiceWorkerTypes.h:
* Source/WebCore/xml/XMLHttpRequest.h:
* Source/WebGPU/WGSL/WGSL.h:
* Source/WebKit/GPUProcess/ShapeDetection/ShapeDetectionObjectHeap.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/WebGPUObjectHeap.h:
* Source/WebKit/NetworkProcess/NetworkLoadChecker.h:
* Source/WebKit/NetworkProcess/cache/NetworkCacheData.h:
* Source/WebKit/Platform/IPC/ArgumentCoders.h:
* Source/WebKit/Platform/IPC/MessageReceiveQueueMap.h:
* Source/WebKit/Platform/cocoa/MediaPlaybackTargetContextSerialized.h:
* Source/WebKit/Scripts/PreferencesTemplates/WebPreferencesStoreDefaultsMap.cpp.erb:
* Source/WebKit/Scripts/webkit/messages.py:
(class_template_headers):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebCompiledContentRuleListData.h:
* Source/WebKit/Shared/WebFoundTextRange.h:
* Source/WebKit/Shared/WebGPU/WebGPUColor.h:
* Source/WebKit/Shared/WebGPU/WebGPUError.h:
* Source/WebKit/Shared/WebGPU/WebGPUExtent3D.h:
* Source/WebKit/Shared/WebGPU/WebGPUImageCopyExternalImage.h:
* Source/WebKit/Shared/WebGPU/WebGPUOrigin2D.h:
* Source/WebKit/Shared/WebGPU/WebGPUOrigin3D.h:
* Source/WebKit/Shared/WebGPU/WebGPURenderPassColorAttachment.h:
* Source/WebKit/Shared/WebGPU/WebGPURenderPassDepthStencilAttachment.h:
* Source/WebKit/UIProcess/API/APIWebAuthenticationPanel.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Automation/SimulatedInputDispatcher.cpp:
* Source/WebKit/UIProcess/Automation/SimulatedInputDispatcher.h:
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/SubFrameSOAuthorizationSession.h:
* Source/WebKit/UIProcess/Cocoa/_WKWarningView.h:
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebAuthentication/WebAuthenticationRequestData.h:
* Source/WebKit/WebKit2Prefix.h:
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferBackendHandle.h:
* Source/WebKitLegacy/mac/DOM/DOMHTMLOptionsCollection.mm:
* Tools/TestWebKitAPI/Tests/WTF/CrossThreadCopierTests.cpp:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/FrameTreeChecks.h:

Canonical link: https://commits.webkit.org/293592@main
